Blog Post: POSH a TLS Compiler that Exploits Program Structure (#80)
sampsyo/cs6120: content/blog/2019-10-23-posh/index.md

data dependencies between threads running on different processor cores.
Broadly, these kinds of features are known as Hardware Transactional Memory (HTM).
At the [end of this article](#hardware-transactional-memory) we've included a brief aside on HTM and its
presence in modern processors for those who are interested.

POSH assumes that hardware has support
for the following features:
 - Inputs to tasks are passed via memory, not registers.
 - Hardware automatically detects conflicting memory reads/writes
   between the main thread and speculative tasks
   and then automatically kills or restarts tasks.
 - The ISA extension has the `spawn` and `commit` primitives
   for starting and ending task execution.

Most papers exploiting HTM rely on a very similar set of assumptions,
and indeed, real HTM extensions have guarantees not unlike those listed here.
The primary difference between these assumptions and reality are empirical limitations
on code and working set size for speculative tasks.

# Sources of Performance Improvements of TLS

The goal for TLS (remember, HTM is the set of hardware features, while TLS

> is a software-level technique that utilizes those features),

Reviewer: ```suggestion
is a software-level technique that utilizes those features)
```

attempt to identify such dependencies (e.g., [instruction scheduling](../instruction-scheduling))
in order to improve performance. However, those optimizations must be
conservative in order to preserve program semantics.
Since TLS compilers can rely on runtime support from the hardware to preserve
correctness, they can aggressively overestimate data independence to
maximize potential parallelism.

```C
...
x = f(a);
y = g(b);
z = h(x,y);
...
```
For example, in the above code snippet, the calls to functions `f` and `g`
can probably be parallelized so that the operands to `h` are available as
soon as possible. However, `f` and `g` may be side-effectful functions that
modify shared memory; TLS using HTM is free to parallelize those two calls
without fear of race conditions on that memory. A normal compiler would have
to prove disjointness of their memory accesses to parallelize them automatically.

The authors point out another, more subtle, benefit to TLS: data prefetching.
Even speculative tasks which violate data dependencies are likely to
access data that will be useful to re-executions of that task.
The authors assume (fairly) that the hardware primitives for squashing tasks

> will not rollback cache state; this implies that squashed tasks can

Reviewer: ```suggestion
will not roll back cache state; this implies that squashed tasks can
```

still prefetch useful data into the cache. Re-executions of the failed
task, or even future tasks may benefit from access to this cached data
and see reduced memory access latency.

<img src="prefetchExample.png" style="width:100%"/>

This diagram from the POSH paper shows how "load times" in
speculative tasks are not totally wasted during task failure.
Executing a speculative load from memory improves the performance of loads in future tasks.

# POSH Phases

The POSH compiler optimization is broken into three phases;
  1) _Task Selection_: Chose speculative tasks based on program structure.
  2) _Spawn Hoisting_: Place task initiation (spawn instructions) as early as possible.
  3) _Task Refinement_: Use dynamic profiling to remove tasks that are unlikely to be beneficial.

The first step is chopping up the program into tasks that will benefit
from being run concurrently.
As you might expect, doing this optimally is NP-hard, so POSH has to resort

> to heuristics.

Reviewer: Maybe briefly state what an "optimal" task decomposition would look like? (That is, state the criterion for optimality.) This will help the reader believe that finding such an optimum would be NP-hard.

> Its primarily heuristic leverages the existing high level program structure;

Reviewer: ```suggestion
Its primary heuristic leverages the existing high level program structure;
```

each subroutine call and loop iteration is considered a candidate task.
The authors justify this with some intuition:
```
All these programmer-generated structures are taken
as hints to delineate code sections with a relatively independent and
sizable amount of work.

> ```

Reviewer: You can use a Markdown block quote for such things:

```md
> All these programmer-generated structures are taken
> as hints to delineate code sections with a relatively independent and
> sizable amount of work.
```

In the _Spawn Hoisting_ phase, POSH tries to move spawn instructions as early
as possible without violating dependencies or changing program behavior.
Spawning tasks early increases opportunities for parallel execution and
prefetching, but there are limits to how far we can move them.
For example, it is not sensible to spawn a thread before the assignment
of its input variables. Neither is it clearly beneficial to
move the spawn instruction outside a conditional
statement since that could result in unnecessary code execution.

The final phase uses profiling and simple syntactic criteria, such as task
size and number of inputs, to remove tasks that probably don't improve performance.
Instead, these pieces of the program are executed straight line.
The profiling methodology is to use test inputs and simulate the parallel
execution of the program with a sequential interpreter, while keeping
track of how many dynamic instructions each task executes and how often
it would have to be squashed by the hardware.

# Evaluation

The overall evaluation methodology is to optimize some of the SPECint benchmarks

> using some configuration of the compiler, and look at statistics (e.g. execution time or memory behavior).

Reviewer: ```suggestion
using some configuration of the compiler, and look at statistics (e.g., execution time or memory behavior).
```

are meant to test the sequential integer performance of CPUs and may have few opportunities for parallelism.
On the other hand, the opposite may be true *or* they may represent a good spread of optimizability.

Given the breakdowns that the authors provide, it seems likely that the most significant
contribution of POSH is its dynamic profiler, which allows their other optimizations
to be aggressively optimistic. While hardware support does prevent TLS from impacting
correctness, it doesn't prevent TLS from being a bad idea. The POSH profiler fills this
gap instead and allows techniques like software value prediction and the structured
program heuristic to be utilized without hurting performance.

If HTM were actually widely usable by general purpose programs, then POSH would likely
be an effective optimization for automatically speeding up sequential code!

# Appendix

### Hardware Transactional Memory

Before transactional memory, hardware support for parallel computing
was limited to synchronization primitives such as [`atomic compare-and-swap`](https://en.wikipedia.org/wiki/Compare-and-swap)

> or [`store-conditional`](https://en.wikipedia.org/wiki/Load-link/store-conditional).

Reviewer: Not sure you actually want these in Markdown code backticks? They don't really seem like code.

### HTM Today

In reality, hardware transactional memory has primarily been a failure
and does not see wide use today.
While Intel theoretically supports these kinds of instructions
with [TSX](https://en.wikipedia.org/wiki/Transactional_Synchronization_Extensions),
numerous bug reports have caused them to [disable it on a number of processors](https://www.anandtech.com/show/8376/intel-disables-tsx-instructions-erratum-found-in-haswell-haswelleep-broadwelly).
Furthermore, the [limitations of TSX](https://blog.ret2.io/2019/06/26/attacking-intel-tsx/)
and other such extensions often make using them impractical, unstable and/or insecure.

However, some low-level code *does* utilize HTM to implement
efficient libraries for high performance computing. In these instances,
developers are targeting very specific architectures with very detailed
models of the processor and memory systems. Since developers in
this domain are already concerned with the finicky details that often
make HTM transactions impractical, HTM does offer utility as a more flexible

> and performant synchronization primitive.

Reviewer: Makes me wonder whether someone should try implementing a SOTA TLS system on TSX. The limiting factor is the L1 size and associativity, of course, but maybe you could get _some_ benefit…

### Software Value Prediction

In TLS, there are some code regions which could be parallelized,
but they involve potentially predictable data dependencies.
For example, in a `while` loop, the iteration condition may not
be known before executing the entire body of the loop and thus parallelism
becomes very limited.

Value prediction transforms this sequential execution into
a potentially parallel one by creating data dependencies between
the original variable and a prediction variable. Value prediction
produces code with the following invariant:

*Let x be some variable in the program, pred(x) is its predicted
value and real(x) is its real value. A TLS task that reads pred(x)

> will be squashed whenever real(x) != pred(x).*

Reviewer: Replace "is" with "be" (twice) for more a more grammatical-sounding first sentence?
